kernel: introduce initial stack randomization

tyrantlizard77 opened the pull request:
This is a component of address space layout randomization that we can
implement even though we have a physical address space.

Support for upward-growing stacks omitted for now, it's not done
currently on any of our current or planned architectures.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: introduce initial stack randomization

This is a component of address space layout randomization that we can
implement even though we have a physical address space.

Support for upward-growing stacks omitted for now, it's not done
currently on any of our current or planned architectures.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

lpereira commented on kernel/thread.c:
```diff
@@ -270,6 +270,16 @@ void _setup_new_thread(struct k_thread *new_thread,
+	/* Don't need to worry about alignment of the size here, _new_thread()
+	 * is required to do it
+	 */
+	stack_size -= sys_rand32_get() % CONFIG_STACK_POINTER_RANDOM;
```
I worry that using the MOD operator here will bias the offset towards a predictable value... specially since sys_rand32_get() does not offer any distribution guarantees, the range of possible random values depends on the sys_rand32_get() implementation, and CONFIG_STACK_POINTER_RANDOM is a user-defined value.

A good way to fix this is to separate the random number generator space into `CONFIG_STACK_POINTER_RANDOM` bins and keep generating numbers if they don't fit in any of those bins.  Went looking for an example code in  a random library and stumbled upon this [Stackoverflow answer that seems pretty decent](https://stackoverflow.com/a/6852396).

This should be fixed at the random subsys level, of course, as there are other places in the kernel where we would at least have some better even distribution guarantees (e.g. stack canary generation) than what sys_rand32_get() offers.

In the meantime, please add a FIXME comment here.

Commit:
arch: add Kconfig for stack growth direction

This was previously just a #define in one header file, but we need
this expressed in Kconfig space in case some feature only works
properly with downward-growing stacks.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented on kernel/Kconfig:
```diff
@@ -556,6 +556,17 @@ config EXECUTE_XOR_WRITE
+	  Randomization by offsetting some random value to a thread's
+	  initial stack pointer upon creation. This hinders some types of
+	  security attacks by making the location of any given stack frame
+	  non-deterministic.
```
done

d3zd3z commented on kernel/thread.c:
```diff
@@ -270,6 +270,19 @@ void _setup_new_thread(struct k_thread *new_thread,
+	 * FIXME: Not the best way to get a random number in a range.
+	 * See #6493
+	 */
+	stack_size -= sys_rand32_get() % CONFIG_STACK_POINTER_RANDOM;
```
I'm a little concerned here about this causing testing to become non-deterministic. If something only slightly overflows its stack, it may work most of the time, only failing rarely, making it hard to debug, especially without overflow checking.

Also, do we need to take into account stack alignment requirements here? Some architectures have strict alignment requirements on the stack, and others will just be less efficient with a misaligned stack.

andrewboie replied:
> testing to become non-deterministic

This isn't on by default, FWIW.

> do we need to take into account stack alignment requirements here

We do not. _new_thread() is required to do any required rounding.

d3zd3z replied:
I don't think the bias is a real concern here, at least as long as sys_rand32_get() has a reasonable distribution. The worst bias is of order `1/ceil(2^32 / CONFIG_STACKPOINTER_RANDOM)`, and given that our stacks are expected to be rather small compared to the address space, I don't think is worth putting in extra code to handle removing the bias.

An extreme example is using 8-bit values, and a modulus of 170, where half of the numbers occur twice as often as others.

d3zd3z commented on tests/kernel/mem_protect/stack_random/src/main.c:
```diff
@@ -0,0 +1,69 @@
+
+void alternate_thread(void)
+{
+	register void *sp __asm__(STACK_REG);
```
Does it work to just take the address of a stack variable? You won't have guarantees of where it is in the stack, but it should definitely move around as the stack is moved.

andrewboie replied:
I'll take a look
